### PR TITLE
Improve dev setup in Makefile

### PR DESCRIPTION
**Pull request by zware**
This is a general cleanup of our Makefile, which takes inspiration from #196 (particularly the 'make help' target) and #195 (particularly the 'contributing' document, which showed me that there was too much global setup required).  It adds some new targets and rewrites others such that the only requirements before running 'make test' are Python (any supported version) and 'make' itself.

### CODE REVIEW

**Review by terencehonles — APPROVED** [review]
LGTM

**Comment by terencehonles** [comment]
It looks like I tested with an existing venv, if I remove it and/or deactivate I get:

```
make lint
python -m venv --clear .venv
make: python: No such file or directory
make: *** [Makefile:70: .venv/bin/poetry] Error 127
```

I realized that error is because Python 2 does not include `venv` the same way and if use Python 3 with `make SYSTEM_PYTHON=python3 lint` I get:

```
pycountry> make SYSTEM_PYTHON=python3 lint
.venv/bin/poetry run pre-commit run --all-files
Command not found: pre-commit
make: *** [Makefile:86: lint] Error 1
```

which is the error I was trying to report.

File: Makefile [hunk 1]
```diff
@@ -1,37 +1,105 @@
...
+## Create a wheel distribution of the project
+wheel: data
 	$(POETRY) build --format=wheel
 
-poetry.lock: pyproject.toml
-	$(POETRY) lock
+.HELP: poetry.lock
+## Ensure that the poetry.lock file is up to date and install all dependencies
+## in the poetry environment
+poetry.lock: pyproject.toml | $(POETRY)
```
Comment (terencehonles) [inline Makefile:65]: I'm not familiar with this syntax, and when removing the venv it doesn't seem like this is installed :thinking: , is this what is meant?

```suggestion
poetry.lock: pyproject.toml $(POETRY)
```

It's now always running these commands, but they seem to be pretty fast

Comment (zware) [reply Makefile:65]: The `|` notes it as an ["ordering only prerequisite"](https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html), meaning (IIUC, makefiles are arcane...) that it basically just checks that the file exists, but doesn't force a regeneration of *this* file if that one has changed.

**Comment by zware** [comment]
> I realized that error is because Python 2 does not include `venv`

Ah, there are still Python 2s kicking around :(.  Will fix `SYSTEM_PYTHON` to default to `python3`.

> if use Python 3 with `make SYSTEM_PYTHON=python3 lint` I get

Here it looks like the problem is that `poetry install` didn't happen.  I'll add a separate marker file for that.

**Pull request merged**